primitives/state-machine/src/in_memory_backend.rs

```diff
@@ -19,134 +19,178 @@
-	backend::{Backend, insert_into_memory_db},
-	stats::UsageInfo,
 };
-use std::{error, fmt, collections::{BTreeMap, HashMap}, marker::PhantomData, ops};
+use std::{collections::{BTreeMap, HashMap}};
 use hash_db::Hasher;
 use sp_trie::{
-	MemoryDB, child_trie_root, empty_child_trie_root, TrieConfiguration, trie_types::Layout,
+	MemoryDB, TrieMut,
+	trie_types::TrieDBMut,
 };
 use codec::Codec;
-use sp_core::storage::{ChildInfo, ChildType, Storage};
+use sp_core::storage::{ChildInfo, Storage};
 
-/// Error impossible.
-// FIXME: use `!` type when stabilized. https://github.com/rust-lang/rust/issues/35121
-#[derive(Debug)]
-pub enum Void {}
-
-impl fmt::Display for Void {
-	fn fmt(&self, _: &mut fmt::Formatter) -> fmt::Result {
-		match *self {}
+/// Insert input pairs into memory db.
+fn insert_into_memory_db<H, I>(mut root: H::Out, mdb: &mut MemoryDB<H>, input: I) -> H::Out
+where
+	H: Hasher,
+	I: IntoIterator<Item=(StorageKey, Option<StorageValue>)>,
+{
+	{
+		let mut trie = if root == Default::default() {
+			TrieDBMut::<H>::new(mdb, &mut root)
+		} else {
+			TrieDBMut::<H>::from_existing(mdb, &mut root).unwrap()
+		};
+		for (key, value) in input {
+			if let Err(e) = match value {
+				Some(value) => {
+					trie.insert(&key, &value)
+				},
+				None => {
+					trie.remove(&key)
+				},
+			}  {
+				panic!("Failed to write to trie: {}", e);
+			}
+		}
+		trie.commit();
 	}
+	root
 }
 
-impl error::Error for Void {
-	fn description(&self) -> &str { "unreachable error" }
-}
-
-/// In-memory backend. Fully recomputes tries each time `as_trie_backend` is called but useful for
-/// tests and proof checking.
-pub struct InMemory<H: Hasher> {
-	inner: HashMap<Option<ChildInfo>, BTreeMap<StorageKey, StorageValue>>,
-	// This field is only needed for returning reference in `as_trie_backend`.
-	trie: Option<TrieBackend<MemoryDB<H>, H>>,
-	_hasher: PhantomData<H>,
+/// Create a new empty instance of in-memory backend.
+pub fn new_in_mem<H: Hasher>() -> TrieBackend<MemoryDB<H>, H>
+where
+	H::Out: Codec + Ord,
+{
+	let db = MemoryDB::default();
+	let mut backend = TrieBackend::new(db, Default::default());
+	backend.insert(std::iter::empty());
+	backend
 }
 
-impl<H: Hasher> std::fmt::Debug for InMemory<H> {
-	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-		write!(f, "InMemory ({} values)", self.inner.len())
+impl<H: Hasher> TrieBackend<MemoryDB<H>, H>
+where
+	H::Out: Codec + Ord,
+{
+	/// Copy the state, with applied updates
+	pub fn update<
+		T: IntoIterator<Item = (Option<ChildInfo>, StorageCollection)>
+	>(
+		&self,
+		changes: T,
+	) -> Self {
+		let mut clone = self.clone();
+		clone.insert(changes);
+		clone
 	}
-}
 
-impl<H: Hasher> Default for InMemory<H> {
-	fn default() -> Self {
-		InMemory {
-			inner: Default::default(),
-			trie: None,
-			_hasher: PhantomData,
+	/// Insert values into backend trie.
+	pub fn insert<
+		T: IntoIterator<Item = (Option<ChildInfo>, StorageCollection)>
+	>(
+		&mut self,
+		changes: T,
+	) {
```

Comment:
```suggestion
	pub fn insert(
		&mut self,
		changes: impl IntoIterator<Item = (Option<ChildInfo>, StorageCollection)>,
	) {
```

Reply:
I'd prefer to stick to the style that's already adopted by the crate. Such changes should be applied with `cargo fix` or some other tool on a per-crate basis.

Reply:
They will never be applied by cargo fix I think, it probably requires compiler context